### PR TITLE
Add user operation event listener archetype

### PR DESCRIPTION
**Pull request by inthirakumaaran**

### CODE REVIEW

File: carbon-user-operation-eventListener-archetype/pom.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>org.wso2.carbon.extension.identity</groupId>
+  <artifactId>user-operation-eventListner-archetype</artifactId>
+  <version>1.0.0-SNAPSHOT</version>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/pom.xml]: change to 5.9.0.1-SNAPSHOT

File: carbon-user-operation-eventListener-archetype/pom.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>org.wso2.carbon.extension.identity</groupId>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/pom.xml]: change to org.wso2.carbon.identity.archetype

File: carbon-user-operation-eventListener-archetype/pom.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  <artifactId>user-operation-eventListner-archetype</artifactId>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/pom.xml]:
```suggestion
  <artifactId>user-operation-event-listener-archetype</artifactId>
```

File: carbon-user-operation-eventListener-archetype/pom.xml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  <packaging>maven-archetype</packaging>
+
+  <name>WSO2 Carbon - User Event Listener Archetype</name>
+
+  <build>
+    <extensions>
+      <extension>
+        <groupId>org.apache.maven.archetype</groupId>
+        <artifactId>archetype-packaging</artifactId>
+        <version>3.0.1</version>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/pom.xml]: use properties

File: carbon-user-operation-eventListener-archetype/setup.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+ * Navigate to `archetypes/carbon-user-operation-eventListener` and run the following command
+         mvn clean install
+
+ * run the following command to create the User operation Event listener
+ 
+   ` mvn archetype:generate
+      -DarchetypeGroupId=org.wso2.carbon.extension.identity
+      -DarchetypeArtifactId=user-operation-eventListner-archetype
+      -DarchetypeVersion=1.0.0-SNAPSHOT
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/setup.md]: use placeholders. Replace this with latest archetype version ..5.9.0.x

File: carbon-user-operation-eventListener-archetype/setup.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      -DgroupId=<sampleID>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/setup.md]: first use place holder then use eg.. don't use sample use generated

File: carbon-user-operation-eventListener-archetype/setup.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+      -DartifactId=<sampeArtifactId>
+      -Dversion=<sampleVersion> `
+       
+
+ * Enter the authenticator name after executing steps 1 & 2. Please enter the use operation event listener prefix
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/setup.md]:
```suggestion
 * Enter the user operation event listener name after executing steps 1 & 2.
```

File: carbon-user-operation-eventListener-archetype/setup.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      -Dversion=<sampleVersion> `
+       
+
+ * Enter the authenticator name after executing steps 1 & 2. Please enter the use operation event listener prefix 
+    `eg:- listener-name : : Sample`
+
+ * Confirm the properties values
+     `Y : : Y`
+     
+ * After making the changes,build your project using `mvn clean install` and
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/setup.md]:
```suggestion
 * After making the changes, build your project using `mvn clean install` and 
```

File: carbon-user-operation-eventListener-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml [hunk 9]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<archetype-descriptor
+        xsi:schemaLocation="http://maven.apache.org/plugins/maven-archetype-plugin/archetype-descriptor/1.0.0 http://maven.apache.org/xsd/archetype-descriptor-1.0.0.xsd"
+        name="pass2"
+        xmlns="http://maven.apache.org/plugins/maven-archetype-plugin/archetype-descriptor/1.0.0"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <fileSets>
+        <fileSet filtered="true" packaged="true" encoding="UTF-8">
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml:8]: check and remove the unnecessary files

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.wso2.eclipse.osgi</groupId>
+            <artifactId>org.eclipse.osgi.services</artifactId>
+            <version>${equinox.osgi.services.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.osgi</groupId>
+            <artifactId>org.eclipse.osgi</artifactId>
+            <version>3.9.1.v20130814-1242</version>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml]: add to properties

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.felix</groupId>
+            <artifactId>org.apache.felix.scr.ds-annotations</artifactId>
+            <version>${org.apache.felix.scr.ds-annotations.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.user.core</artifactId>
+            <version>4.4.3</version>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml]: check this. All other places

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <version>${maven.bundle.plugin.version}</version>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Private-Package>${package}.internal</Private-Package>
+                        <Export-Package>!${package}.internal,
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml]:
```suggestion
                        <Export-Package>
                        !${package}.internal,
```

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                            ${package}.*</Export-Package>
+                        <Import-Package>org.apache.commons.logging.*; version="1.0.4",</Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml:80]:
```suggestion
//this is not the recommended way. Please add the required imports in the Import-Package section and remove the DynamicImport-Package section
                        <DynamicImport-Package>*</DynamicImport-Package>
```

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        <maven.scr.plugin.version>1.22.0</maven.scr.plugin.version>
+        <maven.bundle.plugin.version>2.4.0</maven.bundle.plugin.version>
+        <osgi.framework.imp.pkg.version.range>[1.7.0, 2.0.0)</osgi.framework.imp.pkg.version.range>
+        <osgi.service.component.imp.pkg.version.range>[1.2.0, 2.0.0)</osgi.service.component.imp.pkg.version.range>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <equinox.osgi.services.version>3.5.100.v20160504-1419</equinox.osgi.services.version>
+        <commons-collections.wso2.osgi.version.range>[3.2.0,4.0.0)</commons-collections.wso2.osgi.version.range>
+        <org.apache.felix.scr.ds-annotations.version>1.2.4</org.apache.felix.scr.ds-annotations.version>
+    </properties>
+
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/pom.xml:99]: remove empty lines

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+        //This listener should execute before the IdentityMgtEventListener
+        //Hence the number should be < 1357 (Execution order ID of IdentityMgtEventListener)
+        return 1356;
+    }
+
+    @Override
+    public boolean doPreAuthenticate(String userName, Object credential, UserStoreManager userStoreManager)
+            throws UserStoreException {
+
+        //Edit your c.ode here
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java]:
```suggestion
        //Edit your code here
```

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+import org.wso2.carbon.user.api.Permission;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.common.AbstractUserOperationEventListener;
+
+import java.util.List;
+import java.util.Map;
+
+
+
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java]: Remove spaces

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+public class ${listener-name}UserOperationEventListener extends AbstractUserOperationEventListener {
+
+    //Sample Code
+    private static Log audit = LogFactory.getLog(${listener-name}UserOperationEventListener.class);
+    private static String AUDIT_MESSAGE = "Initiator: %s performed the Action: %s on Target: %s ";
+
+    @Override
+    public int getExecutionOrderId() {
+
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java:26]:
```suggestion
 
 //Todo  implementation required
```

Comment (inthirakumaaran) [reply carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java:26]: DO this for all the places

Comment (inthirakumaaran) [reply carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java:26]: mandatory places -->  implementation is required.
other places --> . implement if required

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+        //Hence the number should be < 1357 (Execution order ID of IdentityMgtEventListener)
+        return 1356;
+    }
+
+    @Override
+    public boolean doPreAuthenticate(String userName, Object credential, UserStoreManager userStoreManager)
+            throws UserStoreException {
+
+        //Edit your c.ode here
+        //Sample Code.
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java:42]: Explain what the methods do.. what kind of logic should code there

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+    @Override
+    public boolean doPostGetUserClaimValues(String userName, String[] claims, String profileName, Map<String,
+            String> claimMap, UserStoreManager storeManager) throws UserStoreException {
+
+        //Sample Code.
+        //audit.info(String.format(AUDIT_MESSAGE, getUser(), "doPostGetUserClaimValues", userName));
+        return true;
+    }
+
+    /**
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/__listener-name__UserOperationEventListener.java:447]: comment this method

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/internal/__listener-name__UserOperationEventListenerDataHolder.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
+#set( $symbol_pound = '#' )
+#set( $symbol_dollar = '$' )
+#set( $symbol_escape = '\' )
+/*
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/internal/__listener-name__UserOperationEventListenerDataHolder.java]: Remove the license from all the templates

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/internal/__listener-name__UserOperationEventListenerServiceComponent.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    protected void activate(ComponentContext context) {
+
+        //register the custom listener as an OSGI service.
+        context.getBundleContext().registerService(UserOperationEventListener.class.getName(),
+                new ${listener-name}UserOperationEventListener(),null);
+        log.info("${listener-name}UserOperationEventListenerServiceComponent bundle activated successfully..");
+    }
+
+    protected void deactivate(ComponentContext context) {
+        if (log.isDebugEnabled()) {
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/internal/__listener-name__UserOperationEventListenerServiceComponent.java:37]: unregister the component

File: carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/internal/__listener-name__UserOperationEventListenerServiceComponent.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+            log.debug("${listener-name}UserOperationEventListenerServiceComponent is deactivated ");
+        }
+    }
+
+    @Reference(name = "user.realm.service.default",
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/main/resources/archetype-resources/src/main/java/internal/__listener-name__UserOperationEventListenerServiceComponent.java:47]: check the formatting

File: carbon-user-operation-eventListener-archetype/src/test/resources/projects/basic/archetype.properties [hunk 23]
```diff
@@ -0,0 +1,6 @@
+#Fri Sep 06 11:22:32 IST 2019
```
Comment (inthirakumaaran) [inline carbon-user-operation-eventListener-archetype/src/test/resources/projects/basic/archetype.properties]: check this file

File: carbon-user-operation-eventListener-archetype/pom.xml (hunk 1 above)
Comment (inthirakumaaran) [reply carbon-user-operation-eventListener-archetype/pom.xml]: ack

**Review by omindu — APPROVED** [review]

**Pull request merged**